README.md
# MIDI::Guitar - Emulating plucked guitar

This module provides basic tools to generated MIDI for plucked string
guitar sounds.

    use MIDI::Guitar;

    my $opus = MIDI::Guitar->new();
    $opus->init( signature => '3/4', tempo => 160 );

    # A pluck pattern.
    my $p1 = $opus->pluck( '1 5:80',
                           '2 2-4:80',
                           '3 4-2:80' );

    # A strum pattern.
    my $s1 = $opus->strum( '1 3/6 6:90 1-5:80' );

    # Some chord shapes.
    my $Am = '0 0 2 2 1 0';
    my $E  = '0 2 2 1 0 0';
    my $C  = '0 3 2 0 1 0';

    $opus->play( $p1 => $Am );
    $opus->play( $p1 => $C  );
    $opus->play( $s1 => $E  );
    $opus->finish( file => 'x.midi' );

You can define play patterns (strum and pluck) and play sequence of
strings (chords) using these patterns. It sounds like a real plucked
instrument, which implies that strings keep sounding until played or
explicitly muted.

You can change tempo, slow down (ritardante) and accelerate.

You can change volume, instantaneously or gradually (de/crescendo).

Timing and volumes can be made varying for a more lively feeling.

It understands ASCII guitar tablature.

## INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

## SUPPORT AND DOCUMENTATION

Development of this module takes place on GitHub:
https://github.com/sciurius/perl-MIDI-Guitar.

You can find documentation for this module with the perldoc command.

    perldoc MIDI::Guitar

Please report any bugs or feature requests using the issue tracker on
GitHub.

## COPYRIGHT AND LICENCE

Copyright (C) 2020 Johan Vromans

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
Revision history for MIDI-Guitar

	Prevent tempo/rit/finish to be called on aux tracks.

0.08	2021-02-16

	Allow complex string forms to implement in-measure chord
	changes.

0.07	2021-02-15

	Allow tracks to be written to separate MIDI files.

	Verify increasing order of pattern elements.

	Explicit finish() is now required.

	Allow play to take a list of string arguments.

	Add "play" option to finish().

	Update and add examples.

0.06	2021-02-12

	Allow play to take an array of patters, and select a random one.

	Use formal names for init values (e.g. signature instead of
	sig, tempo instead of bpm).

	Better argument checking.

	Add identification to the MIDI (text_events).

	Update tests and MANIFEST.

0.05	2020-03-19

	Modified combine into aux.

0.04	2020-03-18

	Added possibilty to combine several opuses into a single.

0.03	2020-03-17

	Added volume(), cresc() and test.

0.02	2020-03-17

	Updated documentation,
	Added some examples.
	Added tests.

0.01    2020-03-16

	First version, released on an unsuspecting world.



Makefile.PL
#! perl

use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME                => 'MIDI::Guitar',
    AUTHOR              => 'Johan Vromans <jv@cpan.org>',
    VERSION_FROM        => 'lib/MIDI/Guitar.pm',
    ABSTRACT_FROM       => 'lib/MIDI/Guitar.pm',
    LICENSE		=> 'perl_5',
    PL_FILES            => {},
    MIN_PERL_VERSION	=> '5.010001',
    PREREQ_PM => {
        'ExtUtils::MakeMaker' => 6.46, # for META_MERGE, LICENSE
        'Test::More'	=> 0,
	'MIDI'		=> 0.83,
    },
    META_MERGE => {
      resources    => {
        repository => {
          type => 'git',
	   web  => 'https://github.com/sciurius/perl-MIDI-Guitar',
	   url  => 'https://github.com/sciurius/perl-MIDI-Guitar.git',
	},
	bugtracker => {
	  web => "https://github.com/sciurius/perl-MIDI-Guitar/issues",
	},
      },
      'meta-spec' => {
        version => '2',
	url     => 'https://metacpan.org/pod/CPAN::Meta::Spec',
      },
    },
);


examples/drums.pl
#!/usr/bin/perl

# Example of playing drums. Really.

use strict;
use warnings;
use utf8;

use FindBin;
use lib "$FindBin::Bin/../lib";
use MIDI::Guitar;

my $opus = MIDI::Guitar->new( signature  => '4/4',
			      tempo      => 80,
			      channel    => 9,
			      instrument => 0, # "Standard Drum Kit",
			      strings =>
			      [ "Low Tom",
				"Low-Mid Tom",
				"High Tom",
				"Ride Cymbal 1",
				"Crash Cymbal 1",
				"Pedal Hi-Hat",
				"Closed Hi-Hat",
				"Acoustic Snare",
				"Acoustic Bass Drum",
			      ],
			      midi => 'drums.midi' );

my $p = $opus->pluck( '1.0  1:90 5:50',
		      '1.5  3:70',
		      '2.0  3:70',
		      '2.5  3:90',
		      '3.0  3:70',
		      '3.5  3:70',
		      '4.0  3:70',
		      '4.5  3:70' );

my $c = "@{[(0) x 9]}";

$opus->play( $p => $c );
$opus->play( $p => $c );

#$opus->tab( 4, xp(<<EOD) );
#AB 0---------------|0---------------|0---------------|0---------------|
#AS ----------------|----------------|----------------|----------------|
#CH ---0--0-0-0-0---|-0-0--0-0-0-0-0-|-0-0--0-0-0-0-0-|-0-0--0-0-0-0-0-|
#PH --------------0-|----------------|----------------|----------------|
#CC -0--------------|----------------|----------------|----------------|
#RC ----------------|----------------|----------------|----------------|
#HT ----------------|----------------|----------------|----------------|
#MT ----------------|----------------|----------------|----------------|
#LT ----------------|----------------|----------------|----------------|
#EOD


# Let the strings sounds decay. 'play' without arguments is a measure
# of nothing.
$opus->play;
$opus->play;

# Finish and write the MIDI file.
# This is not really needed since it will be done implicitly when the
# opus object is destroyed.
# $opus->finish;

sub xp {
    my ( $t ) = @_;
    # Replace leading names by a bar.
    $t =~ s;^.. ;|;igm;
    # Replace anything not |- by -.
    $t =~ s;[^-|\n\r];-;igm;
    return $t
}


examples/drums2.pl
#!/usr/bin/perl

# Example of playing drums. Really.

use strict;
use warnings;
use utf8;

use FindBin;
use lib "$FindBin::Bin/../lib";
use MIDI::Percussion;

my $opus = MIDI::Percussion->new( signature   => '4/4',
				  tempo       => 80,
				  instruments =>
				  [ "Closed Hi-Hat",
				    "Pedal Hi-Hat",
				    "Crash Cymbal 1",
				    "Acoustic Bass Drum",
				  ],
				  midi => 'drums2.midi' );

$opus->tab( <<EOD );
Cl HH |---7--7-7-7-7---|-7-7--7-7-7-7-7-|-7-7--7-7-7-7-7-|-7-7--7-7-7-7-7-|
Pd HH |--------------7-|----------------|----------------|----------------|
Crash |-7--------------|----------------|----------------|----------------|
Kick  |7---------------|7---------------|7---------------|7---------------|
EOD

# Let the strings sounds decay. 'play' without arguments is a measure
# of nothing.
$opus->play;
$opus->play;

# Finish and write the MIDI file.
# This is not really needed since it will be done implicitly when the
# opus object is destroyed.
# $opus->finish;


examples/ocells.pl
#!/usr/bin/perl

# El Cant Dels Ocells by Pau Casals
# https://tabs.ultimate-guitar.com/tab/pau-casals/el-cant-dels-ocells-tabs-1571325

use strict;
use warnings;
use utf8;

use FindBin;
use lib "$FindBin::Bin/../lib";
use MIDI::Guitar;

my $opus = MIDI::Guitar->new( signature => '4/4',
			      tempo     => 70,
			      ticks     => 18*4,
			      midi      => 'ocells.midi' );
sub xp($) {
    my ( $t ) = @_;
    $t =~ s/[-|]$//mg;
    $t =~ s/(.{18})/$1|/g;
    $t =~ s/^/|/mg;
    return $t
}

$opus->tab( 4, xp <<EOD );
-------------------------------------------------------------------------
----------------1-----0------------1-----------3-----------1------2------
----------------2---2-1--------0---2-----0-----2--2--1-----2------0------
--------------0-2-----2--2--3------2--3-----3--0--------3--2---2--2------
-----0--2--3----0-----0------------0-----------------------0------0------
--0----------------------------------------------------------------------
EOD
$opus->tab( 4, xp <<EOD );
-------------------------------------------------------------------------
--------------3------0-----------------------------------1-----2---------
--------------2------0----------0-------------2----1-----2-----0---------
-----0--2--3--0---0--0--3--2--0-2--0----------0----2-----2---2-2---------
--4-----------------------------3-----3--2--3-0--3-2---0-0-----0---------
----------------------------------------------1----0---------------------
EOD
$opus->tab( 4, xp <<EOD );
-----------------------------------------------------------0--5-4-----3--
--------------3-----0--------------------------------0-1-3-1----5-5-6----
----------0---2-----0--------0-----------2----1----2-------2----4--------
----0-2-3---3-0---0-0--3-2-0-2--0--------0----2--2---------2----6--------
--4--------------------------3----3--2-3-0-3--2------------0----0--------
-----------------------------------------1----0--------------------------
EOD
$opus->tab( 4, xp <<EOD );
--0--1--3--1--5--5--------0------------------------3--1--0----0----------
--1-----------6-----9--6--1--5--5--------5--6-6--3-3--------3-1--3--1----
--2-----------7-----------2-----6--6--7-------7----0----------0----------
--2-----------0-----------2-----5-------------7----0----------2----------
--0-----------------------0-----0-------------5---------------3----------
-------------------------------------------------------------------------
EOD
$opus->tab( 4, xp <<EOD );
----------0-----0------------------------3-1-0---0--------------0---5---|
--0-1-3-1-0-----1--5-5------5-6-8-6-6--3-3-----3-1--3-1-0-1-3-1-0---5---|
------2---1---2-2----6--6-7---------7----0-------0----------2---1---5---|
------3---2-----2----5--------------7----0-------2----------3---2---7---|
----------------0----0--------------5------------3------------------7---|
--------------------------------------------------------------------5---|
EOD

# Let the strings sounds decay. 'play' without arguments is a measure
# of nothing.
$opus->play;
$opus->play;

# Finish and write the MIDI file.
# This is not really needed since it will be done implicitly when the
# opus object is destroyed.
# $opus->finish;



examples/stairway.pl
#!/usr/bin/perl

# Example of playing TAB data.

use strict;
use warnings;
use utf8;

use FindBin;
use lib "$FindBin::Bin/../lib";
use MIDI::Guitar;

my $opus = MIDI::Guitar->new( signature => '4/4',
			      tempo     => 80,
			      midi      => 'stairway.midi' );

# The tab is downloaded from one of the many internet sites that have
# all kinds of tabs. Since there are many forms of guitar tabs, and
# our idea of tabs is currently a bit limited, we need to transform
# the downloaded tab into a more suitable format.

sub xp($) {
    my ( $t ) = @_;
    # Replace leading string names by a bar.
    $t =~ s;^[EADGB]-;|-;igm;
    # This tab has a leading - after each bar. Strip it.
    $t =~ s;\|-;|;g;
    # Strip pulls and hammers until we know how to handle.
    $t =~ s;[/^];-;g;
    return $t
}

$opus->tab( 4, xp <<EOD );
E-------5-7-----7-|-8-----8-2-----2-|-0---------0-----|-----------------|
B-----5-----5-----|---5-------3-----|---1---1-----1---|---1-1-----------|
G---5---------5---|-----5-------2---|-----2---------2-|-0-2-2-----------|
D-7-------6-------|-5-------4-------|-3---------------|-----------------|
A-----------------|-----------------|-----------------|-2-0-0---0--/8-7-|
E-----------------|-----------------|-----------------|-----------------|
EOD

$opus->tab( 4, xp <<EOD );
E---------7-----7-|-8-----8-2-----2-|-0---------0-----|-----------------|
B-------5---5-----|---5-------3-----|---1---1-----1---|---1-1-----------|
G-----5-------5---|-----5-------2---|-----2---------2-|-0-2-2-----------|
D---7-----6-------|-5-------4-------|-3---------------|-----------------|
A-0---------------|-----------------|-----------------|-2-0-0-------0-2-|
E-----------------|-----------------|-----------------|-----------------|
EOD

$opus->tab( 4, xp <<EOD );
E-------0-2-----2-|-0-----0---------|---------3-----3-|-3^2-2-2---------|
B-----------3-----|---1-----0-------|-1-----1---0-----|-----3-3---------|
G-----0-------2---|-----2-----2-----|---0---------0---|-----------------|
D---2-----0-------|-3---------------|-----2-----------|-0---0-0---------|
A-3---------------|---------0---0-2-|-3---------------|-------------0-2-|
E-----------------|-----------------|---------3-------|-----------------|
EOD

$opus->tab( 4, xp <<EOD );
E---------2-----2-|-0-----0---------|---------------2-|-0-0-0-----------|
B-------1---3-----|---1-----0-------|-------1-----3---|-1-1-1-----------|
G-----0-------2---|-----2-----2-----|-----0-----2-----|-2-2-2-----------|
D---2-----0-------|-3---------------|---2-----0-------|-3-3-3-----------|
A-3---------------|---------0---0-2-|-3---------------|-----------------|
E-----------------|-----------------|-----------------|-----------------|
EOD

# Let the strings sounds decay. 'play' without arguments is a measure
# of nothing.
$opus->play;
$opus->play;

# Finish and write the MIDI file.
# This is not really needed since it will be done implicitly when the
# opus object is destroyed.
# $opus->finish;



lib/MIDI/Percussion.pm
#! perl

package MIDI::Percussion;

use warnings;
use strict;
use parent qw( MIDI::Guitar );

=head1 NAME

MIDI::Percussion - MIDI Percussion

=cut

our $VERSION = $MIDI::Guitar::VERSION;

=head1 SYNOPSIS

This module provides basic tools to generated MIDI for percussion.

It is derived from MIDI::Guitar, see L<MIDI::Guitar> for features
and behaviour.

    use MIDI::Percussion;

    my $opus = MIDI::Percussion->new( signature   => '4/4',
				      tempo       => 80,
				      instruments =>
				      [ "Closed Hi-Hat",
					"Pedal Hi-Hat",
					"Crash Cymbal 1",
					"Acoustic Bass Drum",
				      ],
				      midi => 'd.midi' );

    $opus->tab( <<EOD );
    Cl HH |---7--7-7-7-7---|-7-7--7-7-7-7-7-|-7-7--7-7-7-7-7-|-7-7--7-7-7-7-7-|
    Pd HH |--------------7-|----------------|----------------|----------------|
    Crash |-7--------------|----------------|----------------|----------------|
    Kick  |7---------------|7---------------|7---------------|7---------------|
    EOD

    $opus->finish;

=cut

use Carp;

=head1 METHODS

=head2 $opus = MIDI::Percussion->new( %args )

Creates a new MIDI::Percussion instance, supplies default values, and
initialises it with the arguments, if any.

Possible arguments are:

=over

=item name

The MIDI track name.

=item signature

Time signature. This should be a fraction where the denominator is a
power of 2, e.g. C<4/4> or C<6/8>.

Handled by the base class. Default is C<4/4>.

=item tempo

Tempo, in beats per minute.

Handled by the base class. Default is C<100>.

=item kit

Drum kit selection. 0 = Standard Drum Kit.

=item instruments

An array reference with instrument names.

There is no default.

Note: There can be any number of instruments.

=item rtime

Handled by the base class. Time randomizer. Suitable values are 0 .. 10.

=item rvolume

Handled by the base class. Volume randomizer. Suitable values are 0 .. 6.

=item lead_in

Lead-in ticks.

Specifies the number of lead-in beats (not bars).

Handled by the base class. Default is 0 (no lead_in).

=item metronome

If nonzero, a metronome tick in included in the MIDI.

Handled by the base class.

=item midi

The name of the MIDI file to be produced.

Handled by the base class.

=back

=cut

sub new {
    my $pkg = shift;
    my $self = bless {} => $pkg;
    my %opts = ( @_ );

    # Defaults.
    my %args = ( name       => "Percussion",
		 channel    => 9,
		 signature  => '4/4',
		 tempo      => 100,
		 instrument => 0,
		 strings => [ "Closed Hi-Hat",
			      "Pedal Hi-Hat",
			      "Crash Cymbal 1",
			      "Acoustic Bass Drum",
			    ],
		 volume  => 1,
	       );

    $args{instrument} = delete $opts{kit} if defined $opts{kit};
    $args{strings} = delete $opts{instruments};
    $self->_init( %args, %opts );
    return $self;
}

=head2 $opus->tab( $measures, $data )

WARNING: Experimental method. API is likely to change.

Plays I<measures> of tablature I<data>.

Support is still very basic, but interestingly functional.

For example:

    $opus->tab( <<EOD );
    Cl HH  |----5-----5-----|--5-------3-----|--1---1-----1---|--1-1-----------|
    Ped HH |--5---------5---|----5-------2---|----2---------2-|0-2-2-----------|
    Crash  |7-------6-------|5-------4-------|3---------------|----------------|
    Kick   |----------------|----------------|----------------|----------------|
    EOD

Note: Each line B<must> contain a series of dashes and numbers between
vertical bars (a measure). Multiple measures may be specified adjacent
to another. Each measure B<should> contain a number of characters equal to
an integer multiple of the beats per measure.

    |5|             5 on one
    |54|            5 on one, 4 on three
    |5444|          5 on one, 4 on two, three, four
    |5-4-4-4-|      same
    |5--4--4--4--|  same
    |5--4--4--444|  same, with triplet on four

The numbers denote the force with which the instrument will be hit.
Zero = mute, 9 = loudest.

The instrument mnemonics are optional and have no relation with the
actual instruments used. The instruments are defined top to bottom by
the C<strings> argument abome.

Returns itself.

=cut

sub tab {
    my ( $self, $tab ) = @_;

    my @args = split( /\s*[\n\r]\s*/, $tab );
    croak("Number of tab lines must be ", scalar(@{ $self->{root} }) )
      unless @args == @{ $self->{root} };

    my $i = 0;
    for ( @args ) {
	if ( /(\|(?:[-\d]+\|)+)$/ ) {
	    $_ = $1;
	}
	else {
	    croak("Tab format error[$i]: $_")
	}
	$i++;
    }

    my $cclock = $self->{clock};

    my $step;
    while ( $args[0] ) {
	my $vel = 100;

	# Randomize clock time and volume.
	my $dv = 0;
	$dv = ( 1 - rand(2) ) * $self->{rvol} if $self->{rvol};
	my $dc = 0;
	$dc = ( 1 - rand(2) ) * $self->{rtime} if $self->{rtime};
	$dc = 0 if $dc + $cclock <= 0;

	my $bar = 0;
	for ( my $s = 0; $s < @{ $self->{root} }; $s++ ) {
	    my $c = substr( $args[$s], 0, 1, '' );
	    if ( $c eq '|' ) {
		if ( $s == 0 ) {
		    if ( $args[0] && $args[0] =~ /^([-\d]+)\|/ ) {
			$step = length($1);
			croak("Strange measure size $step: $args[0]")
			  unless $self->{ticks}/$step == int($self->{ticks}/$step);
			$step = $self->{ticks} / $step * $self->{bpm};
		    }
		    $bar++;
		    $cclock -= $step;
		    next;
		}
		elsif ( $bar ) {
		    next;
		}
		else {
		    delete( $self->{events} ); # prevent destroy finish
		    croak("Tab format error[$s]: $c$args[$s]");
		}
	    }
	    next if $c eq '-';
	    my $str = @{ $self->{root} } - $s - 1;
	    my $note = $self->{root}->[$str];
	    #if ( $self->{sounding}->[$str] ) {
	    #	$self->note( $cclock+$dc, $self->{sounding}->[$str], 0 );
	    #}
	    my $vel = int( $vel * $c/9 );
	    $self->note( $cclock+$dc, $note,
		  $vel > $self->{rvol} ? $vel + $dv : $vel );
	    $self->{sounding}->[$str] = $vel ? $note : 0;
	}
	$cclock += $step;
    }

    $self->{clock} = $cclock;

    return $self;
}

=head1 SEE ALSO

L<MIDI>, L<MIDI::Guitar>.

=head1 COPYRIGHT & LICENSE

Copyright 2020 Johan Vromans, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1; # End of MIDI::Percussion


examples/risingsun.pl
#!/usr/bin/perl

use strict;
use warnings;
use utf8;

use FindBin;
use lib "$FindBin::Bin/../lib";
use MIDI::Guitar;

=for doc

  This example plays the strum pattern made eternally famous 55 years
  ago by the Animals in their version of The House Of The Rising Sun.

  When you play the midi file you'll hear the first 16 bars of the
  song. The rest of the song is just repeating these 16 bars.

=cut

my $opus = MIDI::Guitar->new( signature  => '6/8',
			      tempo      => 240,
			      instrument => 'Electric Guitar(clean)',
			      rtime      =>  10,
			      rvolume    =>   5,
			      midi       => 'risingsun.midi' );

# Chord shapes.
my $Am = '0 0 2 2 1 0';
my $C  = '0 3 2 0 1 0';
my $D  = '- - 0 2 3 2';
my $F  = '1 3 3 2 1 1';
my $E  = '0 2 2 1 0 0';

# Patterns.
# Bass on the A string.
# Note we need to mute the other bass strings to prevent them from
# sounding through the new chord.
# Note that we use imprecise timings so we need pluck, not strum.
my $pA = $opus->pluck( '1.0  5:90  4,6:0',
		       '2.1  4:80',
		       '2.8  3:80',
		       '3.2  2:80',
		       '4.0  1:90',
		       '5.0  2:80',
		       '6.0  3:80' );

# Bass on the D string.
my $pD = $opus->pluck( '1.0  4:90  5,6:0',
		       '2.1  4:80',
		       '2.8  3:80',
		       '3.2  2:80',
		       '4.0  1:90',
		       '5.0  2:80',
		       '6.0  3:80' );

# Bass on the (lower) E string.
my $pE = $opus->pluck( '1.0  6:90  4,5:0',
		       '2.1  4:80',
		       '2.8  3:80',
		       '3.2  2:80',
		       '4.0  1:90',
		       '5.0  2:80',
		       '6.0  3:80' );

# And play...
$opus->play( $pA => $Am );
$opus->play( $pA => $C  );
$opus->play( $pD => $D  );
$opus->play( $pD => $F  );

$opus->play( $pA => $Am );
$opus->play( $pA => $C  );
$opus->play( $pE => $E  );
$opus->play( $pE => $E  );

$opus->play( $pA => $Am );
$opus->play( $pA => $C  );
$opus->play( $pD => $D  );
$opus->play( $pD => $F  );

$opus->play( $pA => $Am );
$opus->play( $pE => $E  );
$opus->play( $pA => $Am );
$opus->play( $pE => $E  );

# Let the strings sounds decay. 'play' without arguments is a measure
# of nothing.
$opus->play;
$opus->play;

# Finish and write the MIDI file.
# This is not really needed since it will be done implicitly when the
# opus object is destroyed.
$opus->finish;

